### PR TITLE
SubjectKeyIdentifier classmethod

### PR DESCRIPTION
**Pull request by reaperhulk**
This class method is needed for creating the SKI extension that needs to be encoded when generating a certificate.

Unfortunately pyasn1 is not really geared towards giving us the data we need so the current solution is very inefficient. I'm not particularly worried about it from a performance perspective, but if there's an easy way to do this more efficiently (not by writing our own ASN.1 parser plz) I'm on board.

Longer term we could also look at the ASN1 lib mentioned in #2162.

refs #2192

### CODE REVIEW

File: docs/x509/reference.rst [hunk 1]
```diff
@@ -1119,6 +1119,21 @@ X.509 Extensions
 
         The binary value of the identifier.
 
+    .. classmethod:: create_from_public_key(public_key)
```
Comment (alex) [inline docs/x509/reference.rst]: Maybe name it `from_public_key`?

File: src/cryptography/x509.py [hunk 2]
```diff
@@ -669,6 +681,29 @@ class SubjectKeyIdentifier(object):
...
+        # pyasn1 really doesn't want to give them to us. To get it we'll
+        # build an integer, hex it, then decode the hex.
+        bits = 0
+        for bit in spki.getComponentByName("subjectPublicKey"):
+            bits = bits << 1 | bit
+
+        # convert the integer to bytes
+        hex_string = '%x' % bits
+        n = len(hex_string)
+        data = binascii.unhexlify(hex_string.zfill(n + (n & 1)))
```
Comment (alex) [inline src/cryptography/x509.py]: Does it maybe make sense to have `public_key.public_bytes(DER, SubjectPublicKey)` (no `Info`)?

Comment (reaperhulk) [inline src/cryptography/x509.py]: We'd want to limit that to DER only and its utility as "the bytes you hash to generate a key identifier" is too limited to be of general benefit in my opinion.

File: docs/x509/reference.rst (hunk 1 above)
Comment (reaperhulk) [inline docs/x509/reference.rst]: renamed.

File: src/cryptography/x509.py [hunk 3]
```diff
@@ -669,6 +681,29 @@ class SubjectKeyIdentifier(object):
...
+        )
+        # the univ.BitString object is a tuple of bits. We need bytes and
+        # pyasn1 really doesn't want to give them to us. To get it we'll
+        # build an integer, hex it, then decode the hex.
+        bits = 0
+        for bit in spki.getComponentByName("subjectPublicKey"):
+            bits = bits << 1 | bit
+
+        # convert the integer to bytes
+        hex_string = '%x' % bits
```
Comment (alex) [inline src/cryptography/x509.py]: This should probably be `hex()`.

File: src/cryptography/x509.py (hunk 2 above)
Comment (alex) [inline src/cryptography/x509.py]: Can you epxlain what algorithm is being implemented, I don't understand what's going on with the hex.

File: src/cryptography/x509.py [hunk 4]
```diff
@@ -669,6 +681,29 @@ class SubjectKeyIdentifier(object):
...
         self._digest = digest
 
+    @classmethod
+    def from_public_key(cls, public_key):
+        # This is a very slow way to do this.
+        serialized = public_key.public_bytes(
+            serialization.Encoding.DER,
+            serialization.PublicFormat.SubjectPublicKeyInfo
+        )
+        spki, remaining = decoder.decode(
```
Comment (alex) [inline src/cryptography/x509.py]: `assert not remainign` or something?

File: src/cryptography/x509.py (hunk 3 above)
Comment (reaperhulk) [inline src/cryptography/x509.py]: We can switch to hex() but then we'll need to [2:] to strip the 0x and .rstrip("L") to get rid of the L in Python 2. That seems uglier, but admittedly `%` is a weird beast in Python with its "not recommended but never going to be removed" status.

If you'd prefer it to swap to `hex(bits)[2:].rstrip("L")` I'll change it.

File: src/cryptography/x509.py (hunk 2 above)
Comment (reaperhulk) [inline src/cryptography/x509.py]: We take the integer, convert to hex, zfill it to make sure it has an even number of characters, then unhexlify it to bytes. It's the reverse of `int.from_bytes(bytes, "big")`

```python
# python3
>>> import binascii
>>> value = 383842938479274659827346598273465298376452398476529038457
>>> hex_string = '%x' % value
>>> n = len(hex_string)
>>> data = binascii.unhexlify(hex_string.zfill(n + (n & 1)))
>>> data
b'\x0f\xa7\x81\xdc5\x1d\xd9\xa0%\x97\x964j\x01\xd8\xf1Z\xad\x93\xb1\xd4\x90\xf4y'
>>> int.from_bytes(data, "big")
383842938479274659827346598273465298376452398476529038457
>>> int.from_bytes(data, "big") == value
True
```

File: docs/x509/reference.rst [hunk 5]
```diff
@@ -1119,6 +1119,21 @@ X.509 Extensions
 
         The binary value of the identifier.
 
+    .. classmethod:: from_public_key(public_key)
+
+        .. versionadded:: 1.0
+
+        Creates a new SubjectKeyIdentifier instance using the public key
+        provided to generate the appropriate digest. This should be the public
+        key that is in the certificate.
```
Comment (alex) [inline docs/x509/reference.rst]: From reading the RFC, it sounds like the SHA1 of the bytes is only one option, is that right?

File: src/cryptography/x509.py (hunk 2 above)
Comment (alex) [inline src/cryptography/x509.py]: Can we factor this logic out into `utils.int_to_bytes` (and use the method on `int` where available)

File: docs/x509/reference.rst (hunk 5 above)
Comment (reaperhulk) [inline docs/x509/reference.rst]: That is correct, although it's the only option I've ever seen used. I'll add some more prose to explain that it's the SHA1 of the subjectPublicKey BIT STRING and link the RFC section

Comment (alex) [inline docs/x509/reference.rst]: Thanks

On Sat, Aug 8, 2015 at 10:49 AM, Paul Kehrer <notifications@github.com>
wrote:

> In docs/x509/reference.rst
> <https://github.com/pyca/cryptography/pull/2204#discussion_r36578919>:
>
> > @@ -1119,6 +1119,21 @@ X.509 Extensions
> >
> >          The binary value of the identifier.
> >
> > +    .. classmethod:: from_public_key(public_key)
> > +
> > +        .. versionadded:: 1.0
> > +
> > +        Creates a new SubjectKeyIdentifier instance using the public key
> > +        provided to generate the appropriate digest. This should be the public
> > +        key that is in the certificate.
>
> That is correct, although it's the only option I've ever seen used. I'll
> add some more prose to explain that it's the SHA1 of the subjectPublicKey
> BIT STRING and link the RFC section
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2204/files#r36578919>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084

File: src/cryptography/x509.py [hunk 6]
```diff
@@ -669,6 +680,27 @@ class SubjectKeyIdentifier(object):
...
+            serialized, asn1Spec=_SubjectPublicKeyInfo()
+        )
+        assert not remaining
+        # the univ.BitString object is a tuple of bits. We need bytes and
+        # pyasn1 really doesn't want to give them to us. To get it we'll
+        # build an integer, hex it, then decode the hex.
```
Comment (alex) [inline src/cryptography/x509.py]: This comment is now a bit innacurate, the hex crap is behind an abstraction now.

File: src/cryptography/utils.py [hunk 7]
```diff
@@ -46,6 +47,18 @@ def int_from_bytes(data, byteorder, signed=False):
         return result
 
 
+if hasattr(int, "to_bytes"):
+    int_to_bytes = int.to_bytes
+else:
+    def int_to_bytes(integer, byteorder, signed=False):
```
Comment (alex) [inline src/cryptography/utils.py]: Eh, fuck, `int.to_bytes` has an API where you need to include the desired length.

**Pull request merged**